c_src/main.c

```diff
@@ -121,6 +137,8 @@ erlfdb_create_future(ErlNifEnv* env, FDBFuture* future, ErlFDBFutureGetter gette
     f->cancelled = false;
     f->cancel_lock = enif_mutex_create("fdb:future_cancel_lock");
 
+    f->future_proc_env_key = st->future_proc_env_key;
```

Comment:
I thought we couldn't share the `future_proc_env_key` when we MAY have concurrent NIF calls from different erlang processes?

Reply:
As far as I understand from the code, key is just an int value representing index in tsd array. It is allocated globally to avoid collisions, so that each key is used for the same purposes in all threads.

Reply:
the common method maybe create a `nif resource` for the `future` but that is expensive for each call.

Reply:
It's just a constant int, seems that passing it through a resource is an overkill 🤔

Reply:
We need to pass it here, because there is no good way to access priv state (`st`) from the callback.

Reply:
My point is I don't think all the NIF function (assume they could be called from different processes) call could share the same `env` thus could not use the same key. 

Because proc_env is per process I believe.

Reply:
>My point is I don't think all the NIF function (assume they could be called from different processes) 
The do not share the same env, they only share the key under which the env is stored for the possible usage down the stack.

While the thread executes the `fdb_future_cancel(future->future);` call, it cannot be moved to another thread (it's nif code) so the process env is always valid.